### PR TITLE
Expunge strtoll(3) and strtol(3)

### PR DESCRIPTION
**Pull request by alejandro-colomar**
Remove all calls to strtoll(3) and strtol(3), and replace them by safer calls like getlong() or getnum().

---

Revisions

<details>
<summary>v3</summary>
v3 changes:

-  Rebase

```
$ git range-diff gh/rm_noneg..gh/strtoll rm_noneg..strtoll 
 1:  dcae312d =  1:  bd966bc5 lib/typetraits.h: Add macros that give information about a type
 2:  3088e310 =  2:  afea5fae src/: Use str2[u]l() instead of atoi(3)
 3:  0d3b5e68 =  3:  617570b1 lib/get_gid.c: get_gid(): Reimplement in terms of a2i()
 4:  aecd2148 =  4:  afa3cebc lib/, libsubid/, po/, src/: get_gid(): Move function to "atoi/getnum.h"
 5:  955fbeec =  5:  a83f9774 lib/: Don't open-code get_gid()
 6:  7e485fa3 =  6:  10ec7b6e lib/get_pid.c: get_pid(): Reimplement in terms of a2i()
 7:  65fd4d27 =  7:  bdf47ee8 lib/: get_pid(): Move function to "atoi/getnum.h"
 8:  3cd8b0ef =  8:  fcfb5bf8 lib/atoi/getnum.[ch]: get_fd(): Add function for parsing a file descriptor from a string
 9:  f22ef57d =  9:  cc310f6b lib/get_pid.c: get_pidfd_from_fd(): Don't open-code get_fd()
10:  72573c2d = 10:  51690f85 src/usermod.c: getulong_range(): Reimplement in terms of a2ul()
11:  a3350c44 = 11:  0de3198b lib/get_uid.c: get_uid(): Reimplement in terms of a2i()
12:  be1e25ff = 12:  5d89bf4f lib/, po/, src/: get_uid(): Move function to "atoi/getnum.h"
13:  17194b35 = 13:  16d30dfe lib/limits.c: setrlimit_value(): Reimplement in terms of a2i()
```
</details>
<details>
<summary>v4</summary>
v4 changes:

-  Fix const-correctness issue, which has been uncovered by the const-generic liba2i macros.

```
$ git range-diff rm_noneg gh/strtoll strtoll 
 1:  bd966bc5 =  1:  bd966bc5 lib/typetraits.h: Add macros that give information about a type
 2:  afea5fae =  2:  afea5fae src/: Use str2[u]l() instead of atoi(3)
 3:  617570b1 =  3:  617570b1 lib/get_gid.c: get_gid(): Reimplement in terms of a2i()
 4:  afa3cebc =  4:  afa3cebc lib/, libsubid/, po/, src/: get_gid(): Move function to "atoi/getnum.h"
 5:  a83f9774 =  5:  a83f9774 lib/: Don't open-code get_gid()
 6:  10ec7b6e =  6:  10ec7b6e lib/get_pid.c: get_pid(): Reimplement in terms of a2i()
 7:  bdf47ee8 =  7:  bdf47ee8 lib/: get_pid(): Move function to "atoi/getnum.h"
 8:  fcfb5bf8 =  8:  fcfb5bf8 lib/atoi/getnum.[ch]: get_fd(): Add function for parsing a file descriptor from a string
 9:  cc310f6b =  9:  cc310f6b lib/get_pid.c: get_pidfd_from_fd(): Don't open-code get_fd()
10:  51690f85 = 10:  51690f85 src/usermod.c: getulong_range(): Reimplement in terms of a2ul()
11:  0de3198b = 11:  0de3198b lib/get_uid.c: get_uid(): Reimplement in terms of a2i()
12:  5d89bf4f = 12:  5d89bf4f lib/, po/, src/: get_uid(): Move function to "atoi/getnum.h"
13:  16d30dfe = 13:  16d30dfe lib/limits.c: setrlimit_value(): Reimplement in terms of a2i()
 -:  -------- > 14:  cff55356 src/usermod.c: Fix const correctness
```
</details>

<details>
<summary>v4b</summary>

-  Rebase

```
$ git range-diff gh/rm_noneg..gh/strtoll rm_noneg..strtoll 
 1:  bd966bc5 =  1:  abac335a lib/typetraits.h: Add macros that give information about a type
 2:  afea5fae =  2:  66a8ac5f src/: Use str2[u]l() instead of atoi(3)
 3:  617570b1 =  3:  10c74161 lib/get_gid.c: get_gid(): Reimplement in terms of a2i()
 4:  afa3cebc =  4:  44c7aea3 lib/, libsubid/, po/, src/: get_gid(): Move function to "atoi/getnum.h"
 5:  a83f9774 =  5:  26d63106 lib/: Don't open-code get_gid()
 6:  10ec7b6e =  6:  84c3bef0 lib/get_pid.c: get_pid(): Reimplement in terms of a2i()
 7:  bdf47ee8 =  7:  5b4afd79 lib/: get_pid(): Move function to "atoi/getnum.h"
 8:  fcfb5bf8 =  8:  3e29da26 lib/atoi/getnum.[ch]: get_fd(): Add function for parsing a file descriptor from a string
 9:  cc310f6b =  9:  53352ba9 lib/get_pid.c: get_pidfd_from_fd(): Don't open-code get_fd()
10:  51690f85 = 10:  f4b116f9 src/usermod.c: getulong_range(): Reimplement in terms of a2ul()
11:  0de3198b = 11:  e24fbb5f lib/get_uid.c: get_uid(): Reimplement in terms of a2i()
12:  5d89bf4f = 12:  d7f1ca08 lib/, po/, src/: get_uid(): Move function to "atoi/getnum.h"
13:  16d30dfe = 13:  412f4c9c lib/limits.c: setrlimit_value(): Reimplement in terms of a2i()
14:  cff55356 = 14:  86734eae src/usermod.c: Fix const correctness
```
</details>

<details>
<summary>v4c</summary>

-  Rebase

```
$ git range-diff abac335a^..gh/strtoll shadow/master..strtoll 
 1:  abac335a =  1:  d06c81f2 lib/typetraits.h: Add macros that give information about a type
 2:  66a8ac5f =  2:  b8e914ff src/: Use str2[u]l() instead of atoi(3)
 3:  10c74161 =  3:  105697f2 lib/get_gid.c: get_gid(): Reimplement in terms of a2i()
 4:  44c7aea3 =  4:  36db6c51 lib/, libsubid/, po/, src/: get_gid(): Move function to "atoi/getnum.h"
 5:  26d63106 =  5:  310b3d80 lib/: Don't open-code get_gid()
 6:  84c3bef0 =  6:  7b9fe7ab lib/get_pid.c: get_pid(): Reimplement in terms of a2i()
 7:  5b4afd79 =  7:  7914f14a lib/: get_pid(): Move function to "atoi/getnum.h"
 8:  3e29da26 =  8:  7d00e662 lib/atoi/getnum.[ch]: get_fd(): Add function for parsing a file descriptor from a string
 9:  53352ba9 =  9:  044396c5 lib/get_pid.c: get_pidfd_from_fd(): Don't open-code get_fd()
10:  f4b116f9 = 10:  0a81ae8e src/usermod.c: getulong_range(): Reimplement in terms of a2ul()
11:  e24fbb5f = 11:  d9e31c73 lib/get_uid.c: get_uid(): Reimplement in terms of a2i()
12:  d7f1ca08 = 12:  53328602 lib/, po/, src/: get_uid(): Move function to "atoi/getnum.h"
13:  412f4c9c = 13:  4d4b4717 lib/limits.c: setrlimit_value(): Reimplement in terms of a2i()
14:  86734eae = 14:  b85ad7a5 src/usermod.c: Fix const correctness
```
</details>

<details>
<summary>v4d</summary>

-  Rebase

```
$ git range-diff gh/master..gh/strtoll shadow/master..strtoll 
 1:  d06c81f2 =  1:  0e49d068 lib/typetraits.h: Add macros that give information about a type
 2:  b8e914ff =  2:  c424c125 src/: Use str2[u]l() instead of atoi(3)
 3:  105697f2 =  3:  2a315a9d lib/get_gid.c: get_gid(): Reimplement in terms of a2i()
 4:  36db6c51 =  4:  e4be326d lib/, libsubid/, po/, src/: get_gid(): Move function to "atoi/getnum.h"
 5:  310b3d80 =  5:  3aaa5e15 lib/: Don't open-code get_gid()
 6:  7b9fe7ab =  6:  f97438b3 lib/get_pid.c: get_pid(): Reimplement in terms of a2i()
 7:  7914f14a =  7:  a66513e3 lib/: get_pid(): Move function to "atoi/getnum.h"
 8:  7d00e662 =  8:  193ca3e0 lib/atoi/getnum.[ch]: get_fd(): Add function for parsing a file descriptor from a string
 9:  044396c5 =  9:  e428f952 lib/get_pid.c: get_pidfd_from_fd(): Don't open-code get_fd()
10:  0a81ae8e = 10:  25191f59 src/usermod.c: getulong_range(): Reimplement in terms of a2ul()
11:  d9e31c73 = 11:  f15b7334 lib/get_uid.c: get_uid(): Reimplement in terms of a2i()
12:  53328602 = 12:  ab95be16 lib/, po/, src/: get_uid(): Move function to "atoi/getnum.h"
13:  4d4b4717 = 13:  eb4c2612 lib/limits.c: setrlimit_value(): Reimplement in terms of a2i()
14:  b85ad7a5 = 14:  8ee58c14 src/usermod.c: Fix const correctness
```
</details>

<details>
<summary>v5</summary>

-  Use multi-line macros, to make it easier to read so many nested parentheses.
-  Remove unnecessary parentheses.
-  Use uid_t.  @hallyn 
-  Test type_max() and type_min(). @hallyn 

```
$ git range-diff alx/master gh/strtoll strtoll 
 1:  0e49d068 !  1:  3a8cadb8 lib/typetraits.h: Add macros that give information about a type
    @@ lib/typetraits.h (new)
     +#include "sizeof.h"
     +
     +
    -+#define is_unsigned(x)  (((typeof(x)) -1) > 1)
    -+#define is_signed(x)    (((typeof(x)) -1) < 1)
    ++#define is_unsigned(x)                                                        \
    ++(                                                                             \
    ++  (typeof(x)) -1 > 1                                                    \
    ++)
     +
    -+#define stype_max(T)    ((T) (((((T) 1 << (WIDTHOF(T) - 2)) - 1) << 1) + 1))
    -+#define utype_max(T)    ((T) -1)
    -+#define type_max(T)     ((T) (is_signed(T) ? stype_max(T) : utype_max(T)))
    -+#define type_min(T)     ((T) ~type_max(T))
    ++#define is_signed(x)                                                          \
    ++(                                                                             \
    ++  (typeof(x)) -1 < 1                                                    \
    ++)
    ++
    ++
    ++#define stype_max(T)                                                          \
    ++(                                                                             \
    ++  (T) (((((T) 1 << (WIDTHOF(T) - 2)) - 1) << 1) + 1)                    \
    ++)
    ++
    ++#define utype_max(T)                                                          \
    ++(                                                                             \
    ++  (T) -1                                                                \
    ++)
    ++
    ++#define type_max(T)                                                           \
    ++(                                                                             \
    ++  (T) (is_signed(T) ? stype_max(T) : utype_max(T))                      \
    ++)
    ++
    ++#define type_min(T)                                                           \
    ++(                                                                             \
    ++  (T) ~type_max(T)                                                      \
    ++)
     +
     +
     +#endif  // include guard
 -:  -------- >  2:  6db29249 tests/unit/test_typetraits.c: Add tests for typetraits.h macros
 2:  c424c125 =  3:  af646349 src/: Use str2[u]l() instead of atoi(3)
 3:  2a315a9d =  4:  6adcee8d lib/get_gid.c: get_gid(): Reimplement in terms of a2i()
 4:  e4be326d =  5:  7441f8d8 lib/, libsubid/, po/, src/: get_gid(): Move function to "atoi/getnum.h"
 5:  3aaa5e15 =  6:  06ff7cc2 lib/: Don't open-code get_gid()
 6:  f97438b3 =  7:  9bff19ee lib/get_pid.c: get_pid(): Reimplement in terms of a2i()
 7:  a66513e3 =  8:  a7e182ff lib/: get_pid(): Move function to "atoi/getnum.h"
 8:  193ca3e0 =  9:  600072eb lib/atoi/getnum.[ch]: get_fd(): Add function for parsing a file descriptor from a string
 9:  e428f952 = 10:  1f52a6ae lib/get_pid.c: get_pidfd_from_fd(): Don't open-code get_fd()
10:  25191f59 = 11:  eeb9d9c1 src/usermod.c: getulong_range(): Reimplement in terms of a2ul()
11:  f15b7334 = 12:  b1729265 lib/get_uid.c: get_uid(): Reimplement in terms of a2i()
12:  ab95be16 = 13:  cd30ec42 lib/, po/, src/: get_uid(): Move function to "atoi/getnum.h"
13:  eb4c2612 = 14:  a63f5ea7 lib/limits.c: setrlimit_value(): Reimplement in terms of a2i()
14:  8ee58c14 = 15:  0368c108 src/usermod.c: Fix const correctness
 -:  -------- > 16:  5aab8329 src/get_subid_owners.c: Use uid_t for holding UIDs (and GIDs)
```
</details>

<details>
<summary>v5b</summary>

-  Most of it has been reviewed by @hallyn 

```
$ git range-diff alx/master gh/strtoll strtoll 
 1:  3a8cadb8 =  1:  3a8cadb8 lib/typetraits.h: Add macros that give information about a type
 2:  6db29249 =  2:  6db29249 tests/unit/test_typetraits.c: Add tests for typetraits.h macros
 3:  af646349 !  3:  b5abb458 src/: Use str2[u]l() instead of atoi(3)
    @@ Commit message
         atoi(3) easily triggers Undefined Behavior.  Replace it by str2[u]l(),
         which are safe from that, and add type safety too.
     
    +    Reviewed-by: "Serge E. Hallyn" <serge@hallyn.com>
         Signed-off-by: Alejandro Colomar <alx@kernel.org>
     
      ## src/free_subid_range.c ##
 4:  6adcee8d !  4:  956fe5c5 lib/get_gid.c: get_gid(): Reimplement in terms of a2i()
    @@ Metadata
      ## Commit message ##
         lib/get_gid.c: get_gid(): Reimplement in terms of a2i()
     
    +    Reviewed-by: "Serge E. Hallyn" <serge@hallyn.com>
         Signed-off-by: Alejandro Colomar <alx@kernel.org>
     
      ## lib/get_gid.c ##
 5:  7441f8d8 !  5:  1180488f lib/, libsubid/, po/, src/: get_gid(): Move function to "atoi/getnum.h"
    @@ Commit message
         Implement it as an inline function, and add restrict and ATTR_STRING()
         and ATTR_ACCESS() as appropriate.
     
    +    Reviewed-by: "Serge E. Hallyn" <serge@hallyn.com>
         Signed-off-by: Alejandro Colomar <alx@kernel.org>
     
      ## lib/Makefile.am ##
 6:  06ff7cc2 !  6:  c8bda919 lib/: Don't open-code get_gid()
    @@ Commit message
     
         These functions were open-coding get_gid().  Use the actual function.
     
    +    Reviewed-by: "Serge E. Hallyn" <serge@hallyn.com>
         Signed-off-by: Alejandro Colomar <alx@kernel.org>
     
      ## lib/getgr_nam_gid.c ##
 7:  9bff19ee !  7:  5d243ef6 lib/get_pid.c: get_pid(): Reimplement in terms of a2i()
    @@ Metadata
      ## Commit message ##
         lib/get_pid.c: get_pid(): Reimplement in terms of a2i()
     
    +    Reviewed-by: "Serge E. Hallyn" <serge@hallyn.com>
         Signed-off-by: Alejandro Colomar <alx@kernel.org>
     
      ## lib/get_pid.c ##
 8:  a7e182ff !  8:  df28e67a lib/: get_pid(): Move function to "atoi/getnum.h"
    @@ Commit message
         Implement it as an inline function, and add restrict and ATTR_STRING()
         and ATTR_ACCESS() as appropriate.
     
    +    Reviewed-by: "Serge E. Hallyn" <serge@hallyn.com>
         Signed-off-by: Alejandro Colomar <alx@kernel.org>
     
      ## lib/atoi/getnum.c ##
 9:  600072eb !  9:  fb2d58e6 lib/atoi/getnum.[ch]: get_fd(): Add function for parsing a file descriptor from a string
    @@ Metadata
      ## Commit message ##
         lib/atoi/getnum.[ch]: get_fd(): Add function for parsing a file descriptor from a string
     
    +    Reviewed-by: "Serge E. Hallyn" <serge@hallyn.com>
         Signed-off-by: Alejandro Colomar <alx@kernel.org>
     
      ## lib/atoi/getnum.c ##
10:  1f52a6ae ! 10:  07a3dc4e lib/get_pid.c: get_pidfd_from_fd(): Don't open-code get_fd()
    @@ Metadata
      ## Commit message ##
         lib/get_pid.c: get_pidfd_from_fd(): Don't open-code get_fd()
     
    +    Reviewed-by: "Serge E. Hallyn" <serge@hallyn.com>
         Signed-off-by: Alejandro Colomar <alx@kernel.org>
     
      ## lib/get_pid.c ##
11:  eeb9d9c1 ! 11:  26e759c8 src/usermod.c: getulong_range(): Reimplement in terms of a2ul()
    @@ Metadata
      ## Commit message ##
         src/usermod.c: getulong_range(): Reimplement in terms of a2ul()
     
    +    Reviewed-by: "Serge E. Hallyn" <serge@hallyn.com>
         Signed-off-by: Alejandro Colomar <alx@kernel.org>
     
      ## src/usermod.c ##
12:  b1729265 ! 12:  753192f2 lib/get_uid.c: get_uid(): Reimplement in terms of a2i()
    @@ Metadata
      ## Commit message ##
         lib/get_uid.c: get_uid(): Reimplement in terms of a2i()
     
    +    Reviewed-by: "Serge E. Hallyn" <serge@hallyn.com>
         Signed-off-by: Alejandro Colomar <alx@kernel.org>
     
      ## lib/get_uid.c ##
13:  cd30ec42 ! 13:  824f6b3c lib/, po/, src/: get_uid(): Move function to "atoi/getnum.h"
    @@ Commit message
         Implement it as an inline function, and add restrict and ATTR_STRING()
         and ATTR_ACCESS() as appropriate.
     
    +    Reviewed-by: "Serge E. Hallyn" <serge@hallyn.com>
         Signed-off-by: Alejandro Colomar <alx@kernel.org>
     
      ## lib/Makefile.am ##
14:  a63f5ea7 ! 14:  9e9b28c0 lib/limits.c: setrlimit_value(): Reimplement in terms of a2i()
    @@ Metadata
      ## Commit message ##
         lib/limits.c: setrlimit_value(): Reimplement in terms of a2i()
     
    +    Reviewed-by: "Serge E. Hallyn" <serge@hallyn.com>
         Signed-off-by: Alejandro Colomar <alx@kernel.org>
     
      ## lib/limits.c ##
15:  0368c108 ! 15:  015087c8 src/usermod.c: Fix const correctness
    @@ Commit message
         Now that we use liba2i's const-generic macros, we can (and must) use a
         'const char **' endp where the input string is 'const char *'.
     
    +    Reviewed-by: "Serge E. Hallyn" <serge@hallyn.com>
         Signed-off-by: Alejandro Colomar <alx@kernel.org>
     
      ## src/usermod.c ##
16:  5aab8329 = 16:  ee723b59 src/get_subid_owners.c: Use uid_t for holding UIDs (and GIDs)
```
</details>

### CODE REVIEW

**Comment by alejandro-colomar** [comment]
v2 changes:

-  Rebase to master

```
$ git range-diff gh/rm_noneg..gh/strtoll rm_noneg..strtoll 
 1:  d987cc23 !  1:  7a754763 lib/typetraits.h: Add macros that give information about a type
    @@ Commit message
     
      ## lib/Makefile.am ##
     @@ lib/Makefile.am: libshadow_la_SOURCES = \
    -   subordinateio.c \
    -   sulog.c \
    +   time/day_to_str.c \
    +   time/day_to_str.h \
        ttytype.c \
     +  typetraits.h \
        tz.c \
 2:  04c3c1ff =  2:  35a75a2f src/: Use str2[u]l() instead of atoi(3)
 3:  901bf98e =  3:  2a5df6b4 lib/get_gid.c: get_gid(): Reimplement in terms of a2i()
 4:  15664d4c !  4:  0245884e lib/, libsubid/, po/, src/: get_gid(): Move function to "atoi/getnum.h"
    @@ lib/prototypes.h: extern int find_new_sub_gids (gid_t *range_start, unsigned lon
     
      ## lib/sgetgrent.c ##
     @@
    - #include <grp.h>
    + #include <string.h>
      
      #include "alloc.h"
     +#include "atoi/getnum.h"
    @@ lib/sgetgrent.c
     
      ## lib/sgetpwent.c ##
     @@
    - #include "defines.h"
    - #include <stdio.h>
      #include <pwd.h>
    -+
    + #include <string.h>
    + 
     +#include "atoi/getnum.h"
    + #include "defines.h"
      #include "prototypes.h"
      #include "shadowlog_internal.h"
      
 5:  e6d47f2c =  5:  637a8484 lib/: Don't open-code get_gid()
 6:  59b2f005 =  6:  e5b31b34 lib/get_pid.c: get_pid(): Reimplement in terms of a2i()
 7:  23fd5ac6 =  7:  6bd34d8c lib/: get_pid(): Move function to "atoi/getnum.h"
 8:  35b3c86c =  8:  e908ef42 lib/atoi/getnum.[ch]: get_fd(): Add function for parsing a file descriptor from a string
 9:  4e0e9bf1 =  9:  05bd8850 lib/get_pid.c: get_pidfd_from_fd(): Don't open-code get_fd()
10:  8a4f7943 = 10:  d8fa7301 src/usermod.c: getulong_range(): Reimplement in terms of a2ul()
11:  330dd287 = 11:  ae87a7e9 lib/get_uid.c: get_uid(): Reimplement in terms of a2i()
12:  c77cfdab = 12:  27692802 lib/, po/, src/: get_uid(): Move function to "atoi/getnum.h"
13:  68ba7031 = 13:  a0b322e2 lib/limits.c: setrlimit_value(): Reimplement in terms of a2i()
```

**Comment by alejandro-colomar** [comment]
v2c changes:

-  Rebase on master

```
$ git range-diff gh/rm_noneg..gh/strtoll rm_noneg..strtoll 
 1:  6870a0e0 =  1:  8f9ff452 lib/typetraits.h: Add macros that give information about a type
 2:  ebba199f =  2:  413993b3 src/: Use str2[u]l() instead of atoi(3)
 3:  a2122723 =  3:  4c2159dd lib/get_gid.c: get_gid(): Reimplement in terms of a2i()
 4:  6eae04d4 =  4:  dee66fc3 lib/, libsubid/, po/, src/: get_gid(): Move function to "atoi/getnum.h"
 5:  debb96f4 =  5:  c37c2ae4 lib/: Don't open-code get_gid()
 6:  82a992d3 =  6:  8dcd0354 lib/get_pid.c: get_pid(): Reimplement in terms of a2i()
 7:  58f28b8e =  7:  fa78d95c lib/: get_pid(): Move function to "atoi/getnum.h"
 8:  5eb9fc41 =  8:  a5bb3e22 lib/atoi/getnum.[ch]: get_fd(): Add function for parsing a file descriptor from a string
 9:  faa7cce1 =  9:  c7273620 lib/get_pid.c: get_pidfd_from_fd(): Don't open-code get_fd()
10:  14ccc58b = 10:  2def551a src/usermod.c: getulong_range(): Reimplement in terms of a2ul()
11:  fad7d7bb = 11:  a41871ee lib/get_uid.c: get_uid(): Reimplement in terms of a2i()
12:  30d634ba = 12:  5f862873 lib/, po/, src/: get_uid(): Move function to "atoi/getnum.h"
13:  97d5ce57 = 13:  0247134d lib/limits.c: setrlimit_value(): Reimplement in terms of a2i()
```

**Comment by alejandro-colomar** [comment]
v2d changes:

-  Rebase on master

```
$ git range-diff gh/rm_noneg..gh/strtoll rm_noneg..strtoll 
 1:  8f9ff452 =  1:  932fa96d lib/typetraits.h: Add macros that give information about a type
 2:  413993b3 =  2:  97471e54 src/: Use str2[u]l() instead of atoi(3)
 3:  4c2159dd =  3:  a5546d76 lib/get_gid.c: get_gid(): Reimplement in terms of a2i()
 4:  dee66fc3 =  4:  8f8095bc lib/, libsubid/, po/, src/: get_gid(): Move function to "atoi/getnum.h"
 5:  c37c2ae4 =  5:  6ad6667d lib/: Don't open-code get_gid()
 6:  8dcd0354 =  6:  cc67840f lib/get_pid.c: get_pid(): Reimplement in terms of a2i()
 7:  fa78d95c =  7:  dd1413cf lib/: get_pid(): Move function to "atoi/getnum.h"
 8:  a5bb3e22 =  8:  e5ddd6cc lib/atoi/getnum.[ch]: get_fd(): Add function for parsing a file descriptor from a string
 9:  c7273620 =  9:  d66e280a lib/get_pid.c: get_pidfd_from_fd(): Don't open-code get_fd()
10:  2def551a = 10:  21ee50c4 src/usermod.c: getulong_range(): Reimplement in terms of a2ul()
11:  a41871ee = 11:  11c0777f lib/get_uid.c: get_uid(): Reimplement in terms of a2i()
12:  5f862873 = 12:  fd9f2033 lib/, po/, src/: get_uid(): Move function to "atoi/getnum.h"
13:  0247134d = 13:  ae1fd8a2 lib/limits.c: setrlimit_value(): Reimplement in terms of a2i()
```

**Comment by alejandro-colomar** [comment]
v2e changes:

- Rebase

```
$ git range-diff gh/rm_noneg..gh/strtoll rm_noneg..strtoll 
 1:  932fa96d =  1:  cc6c9099 lib/typetraits.h: Add macros that give information about a type
 2:  97471e54 =  2:  ceb9cf25 src/: Use str2[u]l() instead of atoi(3)
 3:  a5546d76 =  3:  80b2f03b lib/get_gid.c: get_gid(): Reimplement in terms of a2i()
 4:  8f8095bc =  4:  f3b68d0f lib/, libsubid/, po/, src/: get_gid(): Move function to "atoi/getnum.h"
 5:  6ad6667d =  5:  73828d71 lib/: Don't open-code get_gid()
 6:  cc67840f =  6:  5c1f3269 lib/get_pid.c: get_pid(): Reimplement in terms of a2i()
 7:  dd1413cf =  7:  9a8e8959 lib/: get_pid(): Move function to "atoi/getnum.h"
 8:  e5ddd6cc =  8:  9c272ae7 lib/atoi/getnum.[ch]: get_fd(): Add function for parsing a file descriptor from a string
 9:  d66e280a =  9:  a079abe2 lib/get_pid.c: get_pidfd_from_fd(): Don't open-code get_fd()
10:  21ee50c4 = 10:  6b4a6d42 src/usermod.c: getulong_range(): Reimplement in terms of a2ul()
11:  11c0777f = 11:  98676974 lib/get_uid.c: get_uid(): Reimplement in terms of a2i()
12:  fd9f2033 = 12:  995d0b87 lib/, po/, src/: get_uid(): Move function to "atoi/getnum.h"
13:  ae1fd8a2 = 13:  1125c9a7 lib/limits.c: setrlimit_value(): Reimplement in terms of a2i()
```

**Comment by alejandro-colomar** [comment]
v2f changes:

-  Rebase

```
$ git range-diff gh/rm_noneg..gh/strtoll rm_noneg..strtoll 
 1:  cc6c9099 =  1:  dcae312d lib/typetraits.h: Add macros that give information about a type
 2:  ceb9cf25 =  2:  3088e310 src/: Use str2[u]l() instead of atoi(3)
 3:  80b2f03b =  3:  0d3b5e68 lib/get_gid.c: get_gid(): Reimplement in terms of a2i()
 4:  f3b68d0f =  4:  aecd2148 lib/, libsubid/, po/, src/: get_gid(): Move function to "atoi/getnum.h"
 5:  73828d71 =  5:  955fbeec lib/: Don't open-code get_gid()
 6:  5c1f3269 =  6:  7e485fa3 lib/get_pid.c: get_pid(): Reimplement in terms of a2i()
 7:  9a8e8959 =  7:  65fd4d27 lib/: get_pid(): Move function to "atoi/getnum.h"
 8:  9c272ae7 =  8:  3cd8b0ef lib/atoi/getnum.[ch]: get_fd(): Add function for parsing a file descriptor from a string
 9:  a079abe2 =  9:  f22ef57d lib/get_pid.c: get_pidfd_from_fd(): Don't open-code get_fd()
10:  6b4a6d42 = 10:  72573c2d src/usermod.c: getulong_range(): Reimplement in terms of a2ul()
11:  98676974 = 11:  a3350c44 lib/get_uid.c: get_uid(): Reimplement in terms of a2i()
12:  995d0b87 = 12:  be1e25ff lib/, po/, src/: get_uid(): Move function to "atoi/getnum.h"
13:  1125c9a7 = 13:  17194b35 lib/limits.c: setrlimit_value(): Reimplement in terms of a2i()
```

**Comment by alejandro-colomar** [comment]
It seems the const-generic macros have exposed some const correctness violations.  They are probably nothing to worry, but I'll fix them.

**Comment by alejandro-colomar** [comment]
This is ready for review, @hallyn , @ikerexxe .  :)

File: lib/typetraits.h [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+
+#include <config.h>
+
+#include "sizeof.h"
+
+
+#define is_unsigned(x)  (((typeof(x)) -1) > 1)
+#define is_signed(x)    (((typeof(x)) -1) < 1)
+
+#define stype_max(T)    ((T) (((((T) 1 << (WIDTHOF(T) - 2)) - 1) << 1) + 1))
```
Comment (hallyn) [inline lib/typetraits.h:17]: Call me chicken, but could you add a simple testcase for just one example type, enforcing that stype_max and utype_max do the right thing for, say, an 8 bit type?

File: src/get_subid_owners.c [hunk 2]
```diff
@@ -18,20 +23,24 @@ static void usage(void)
...
+	if (argc == 3 && strcmp(argv[1], "-g") == 0) {
+		str2sl(&l, argv[2]);
+		n = subid_get_gid_owners(l, &uids);
+	} else if (argc == 2 && strcmp(argv[1], "-h") == 0) {
 		usage();
-	else
-		n = subid_get_uid_owners(atoi(argv[1]), &uids);
+	} else {
+		str2sl(&l, argv[1]);
+		n = subid_get_uid_owners(l, &uids);
```
Comment (hallyn) [inline src/get_subid_owners.c:42]: Hm, you're obviously not introducing this, but technically I guess l should be unsigned long?  Going by 'typedef __kernel_uid32_t        uid_t;' in the kernel.

Comment (alejandro-colomar) [reply src/get_subid_owners.c:42]: We can use `uid_t`.  That is:

```c
str2i(uid_t, &u, argv[1]);
```

**Review by hallyn — APPROVED** [review]
Thank you, this is great.

Feel free to merge as is, or if you want to add a testcase for the type_min/max, that's great too.

File: lib/typetraits.h (hunk 1 above)
Comment (alejandro-colomar) [reply lib/typetraits.h:17]: I'm preparing a test.

**Comment by alejandro-colomar** [comment]
On Sat, Jun 29, 2024 at 10:16:16AM GMT, Serge Hallyn wrote:
> @hallyn approved this pull request.
> 
> Thank you, this is great.
> 
> Feel free to merge as is, or if you want to add a testcase for the type_min/max, that's great too.

I've added your suggestions, plus some readability improvements.
I'll add your reviewed-by for all commits, if you confirm.

Thanks!

Have a lovely day!
Alex

> -- 
> Reply to this email directly or view it on GitHub:
> https://github.com/shadow-maint/shadow/pull/896#pullrequestreview-2149735495
> You are receiving this because you authored the thread.
> 
> Message ID: ***@***.***>

-- 
<https://www.alejandro-colomar.es/>

**Comment by hallyn** [comment]
Awesome - thank you.

**Pull request merged**